pip/pip-272.md

```diff
@@ -0,0 +1,214 @@
+* This section is the "Why" of your proposal.
+-->
+
+Pulsar functions don't support configuring the `StateStoreProviderImplClass` with extra configurations, although there is a `Map<String, Object> config` parameter in the `StateStoreProvider#init` method, it's initialized as a map with only one field `stateStorageServiceUrl` in `javaInstanceRunnable`
+
+see:
+
+https://github.com/apache/pulsar/blob/master/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L366-L369
+
+Then when using a custom `StateStoreProvider`, it cannot get other configurations from the `init` method.
+
+# Goals
+
+## In Scope
+
+<!--
+What this PIP intend to achieve once It's integrated into Pulsar.
+Why does it benefit Pulsar.
+-->
+
+Make `StateStoreProvider` "truly" configurable, custom `StateStoreProvider` can be configured more easily and explicitly.
+
+## Out of Scope
+
+<!--
+Describe what you have decided to keep out of scope, perhaps left for a different PIP/s.
+-->
+
+
+# High Level Design
+
+<!--
+Describe the design of your solution in *high level*.
+Describe the solution end to end, from a birds-eye view.
+Don't go into implementation details in this section.
+
+I should be able to finish reading from beginning of the PIP to here (including) and understand the feature and 
+how you intend to solve it, end to end.
+
+DON'T
+* Avoid code snippets, unless it's essential to explain your intent.
+-->
+
+Make `StateStoreProvider` configurable using custom configurations.
+
+# Detailed Design
+
+## Design & Implementation Details
+
+<!--
+This is the section where you dive into the details. It can be:
+* Concrete class names and their roles and responsibility, including methods.
+* Code snippets of existing code.
+* Interface names and its methods.
+* ...
+-->
+
+1. Add a `stateStorageConfig` to the `WorkerConfig`
+2. In all three runtime factories, pass this config to the created runtime
+3. Add a new cli argument to `JavaInstanceStarter` and `LocalRunner` so process&k8s runtime can pass state related config to them
+4. Add new state config parameter to `JavaInstanceRunnable` to accept the state provider related configurations
+
+## Public-facing Changes
+
+<!--
+Describe the additions you plan to make for each public facing component. 
+Remove the sections you are not changing.
+Clearly mark any changes which are BREAKING backward compatability.
+-->
+
+### Public API
+<!--
+When adding a new endpoint to the REST API, please make sure to document the following:
+
+* path
+* query parameters
+* HTTP body parameters, usually as JSON.
+* Response codes, and for each what they mean.
+  For each response code, please include a detailed description of the response body JSON, specifying each field and what it means.
+  This is the place to document the errors.
+-->
+
+### Binary protocol
+
+### Configuration
+
+- A new `map<String, Object>` field like `stateStorageConfig` is added to `WorkerConfig`
```

Comment:
Not exactly.
I guess you plan to add a config key named `stateStorageConfig`, but how will users enter `map<String, Object>` - don't assume they are java developers and even if they do they will don't know what to put there.

Now I presume the value type is single line JSON String?

Reply:
No, the `WorkerConfig` is a Java class, so this field is a `map<String, Object>` instead of a JSON string, there are already other `map<String, Object>` fields in the `WorkerConfig`, such as `functionRuntimeFactoryConfigs`, `runtimeCustomizerConfig` and `functionsWorkerServiceCustomConfigs`.

On users(operator) side, they just need to edit the `functions_worker.yml` to provide the `stateStorageConfig`, which will be parsed and loaded as the `WorkerConfig`; an example looks like the below:

```yaml
stateStorageConfig:
    username: xxx
    password: xxx
```

Reply:
Sure, I will add the info about `function_worker.yml` in the background